File: src/main/java/archives/tater/omnicrossbow/HoneySlickBlock.java
package archives.tater.omnicrossbow;

import archives.tater.omnicrossbow.mixin.HoneyBlockInvoker;
import net.minecraft.block.*;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.particle.ParticleTypes;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.EnumProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.util.function.BooleanBiFunction;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Vec3d;
import net.minecraft.util.math.random.Random;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import net.minecraft.world.WorldAccess;
import net.minecraft.world.WorldView;

import java.util.Arrays;
import java.util.Map;
import java.util.stream.Collectors;

@SuppressWarnings("deprecation")
public class HoneySlickBlock extends HoneyBlock {
    public static final EnumProperty<Direction> FACING = Properties.FACING;

    public static final Map<Direction, VoxelShape> FACE_SHAPES = Arrays.stream(Direction.values()).collect(Collectors.toUnmodifiableMap(
            direction -> direction,
            direction -> directionalLayer(direction, 4)));
    private static final Map<Direction, VoxelShape> SUPPORT_SHAPES = Arrays.stream(Direction.values()).collect(Collectors.toUnmodifiableMap(
            direction -> direction,
            direction -> directionalLayer(direction.getOpposite(), 1)));

    public HoneySlickBlock(Settings settings) {
        super(settings);
        setDefaultState(getDefaultState().with(FACING, Direction.DOWN));
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        builder.add(FACING);
    }

    @Override
    public VoxelShape getOutlineShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        return getShape(state);
    }

    public VoxelShape getShape(BlockState state) {
        return FACE_SHAPES.get(state.get(FACING));
    }

    @Override
    public VoxelShape getCollisionShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        return VoxelShapes.empty();
    }

    @Override
    public void onEntityCollision(BlockState state, World world, BlockPos pos, Entity entity) {
        if (entity.getBoundingBox().intersects(getShape(state).offset(pos.getX(), pos.getY(), pos.getZ()).getBoundingBox())) { // This only works for simple voxelshapes
            var length = entity instanceof PlayerEntity ? entity.getPos().subtract(entity.prevX, entity.prevY, entity.prevZ).length() : entity.getVelocity().length();
            if (length >= 0.05)
                ((HoneyBlockInvoker) this).invokeAddCollisionEffects(world, entity);
            if (!entity.isOnGround() && entity.getVelocity().y < -0.08) {
                ((HoneyBlockInvoker) this).invokeUpdateSlidingVelocity(entity);
            } else {
                entity.slowMovement(state, new Vec3d(0.9F, 1.5, 0.9F));
            }
        }
    }

    @Override
    public boolean isSideInvisible(BlockState state, BlockState stateFrom, Direction direction) {
        return stateFrom.isOf(this) && state.get(FACING) == stateFrom.get(FACING);
    }

    @Override
    public boolean canPlaceAt(BlockState state, WorldView world, BlockPos pos) {
        var direction = state.get(FACING);
        var basePos = pos.offset(direction);
        return VoxelShapes.matchesAnywhere(world.getBlockState(basePos).getSidesShape(world, basePos).getFace(direction.getOpposite()), SUPPORT_SHAPES.get(direction), BooleanBiFunction.AND);
    }

    @Override
    public BlockState getStateForNeighborUpdate(
            BlockState state, Direction direction, BlockState neighborState, WorldAccess world, BlockPos pos, BlockPos neighborPos
    ) {
        return direction == state.get(FACING) && !state.canPlaceAt(world, pos)
                ? Blocks.AIR.getDefaultState()
                : super.getStateForNeighborUpdate(state, direction, neighborState, world, pos, neighborPos);
    }

    @Override
    public void randomDisplayTick(BlockState state, World world, BlockPos pos, Random random) {
        if (!world.getBlockState(pos.down()).isSideSolid(world, pos.down(), Direction.DOWN, SideShapeType.FULL) && random.nextInt(16) == 0) {
            var box = state.getOutlineShape(world, pos).getBoundingBox();
            world.addParticle(ParticleTypes.DRIPPING_HONEY, box.getXLength() * random.nextDouble() + box.minX + pos.getX(), box.minY + pos.getY() - 0.05, box.getZLength() * random.nextDouble() + box.minZ + pos.getZ(), 0, 0, 0);
        }
    }

    private static VoxelShape directionalLayer(Direction direction, int thickness) {
        return createCuboidShape(
                direction == Direction.EAST ? 16 - thickness : 0,
                direction == Direction.UP ? 16 - thickness : 0,
                direction == Direction.SOUTH ? 16 - thickness : 0,
                direction == Direction.WEST ? thickness : 16,
                direction == Direction.DOWN ? thickness : 16,
                direction == Direction.NORTH ? thickness : 16
        );
    }

}

File: src/main/kotlin/archives/tater/omnicrossbow/block/HoneySlickBlock.kt
package archives.tater.omnicrossbow.block

import archives.tater.omnicrossbow.mixin.HoneyBlockInvoker
import archives.tater.omnicrossbow.util.get
import archives.tater.omnicrossbow.util.minus
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.particles.ParticleTypes
import net.minecraft.server.level.ServerPlayer
import net.minecraft.util.RandomSource
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.InsideBlockEffectApplier
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.ScheduledTickAccess
import net.minecraft.world.level.block.*
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.properties.EnumProperty
import net.minecraft.world.phys.Vec3
import net.minecraft.world.phys.shapes.BooleanOp
import net.minecraft.world.phys.shapes.CollisionContext
import net.minecraft.world.phys.shapes.Shapes
import net.minecraft.world.phys.shapes.VoxelShape


class HoneySlickBlock(properties: Properties) : HoneyBlock(properties) {
    init {
        registerDefaultState(stateDefinition.any().setValue(FACING, Direction.UP))
    }

    override fun createBlockStateDefinition(builder: StateDefinition.Builder<Block, BlockState>) {
        super.createBlockStateDefinition(builder)
        builder.add(FACING)
    }

    override fun getShape(
        state: BlockState,
        level: BlockGetter,
        pos: BlockPos,
        context: CollisionContext
    ): VoxelShape = FACE_SHAPES[state[FACING]]!!

    override fun getCollisionShape(
        state: BlockState,
        level: BlockGetter,
        pos: BlockPos,
        context: CollisionContext
    ): VoxelShape = Shapes.empty()

    override fun getEntityInsideCollisionShape(
        state: BlockState,
        level: BlockGetter,
        pos: BlockPos,
        entity: Entity
    ): VoxelShape = getShape(state,  level, pos, CollisionContext.of(entity))

    override fun entityInside(
        state: BlockState,
        level: Level,
        pos: BlockPos,
        entity: Entity,
        effectApplier: InsideBlockEffectApplier,
        isPrecise: Boolean
    ) {
        @Suppress("CAST_NEVER_SUCCEEDS")
        this as HoneyBlockInvoker

        val movement = if (entity is ServerPlayer) entity.knownMovement else entity.position() - entity.oldPosition()

        if (movement.length() > 0)
            invokeMaybeDoSlideEffects(level, entity)

        if (!entity.onGround() && movement.y < -0.08) {
            invokeDoSlideMovement(entity)
        } else
            entity.makeStuckInBlock(state, Vec3(0.9, 1.5, 0.9))
    }

    override fun skipRendering(state: BlockState, neighborState: BlockState, direction: Direction): Boolean =
        super.skipRendering(state, neighborState, direction) && state[FACING] == neighborState[FACING]

    override fun getStateForPlacement(context: BlockPlaceContext): BlockState =
        defaultBlockState().setValue(FACING, context.clickedFace)

    override fun canSurvive(state: BlockState, level: LevelReader, pos: BlockPos): Boolean {
        val direction = state[FACING]
        val basePos = pos.relative(direction.opposite)
        return Shapes.joinIsNotEmpty(level[basePos].getBlockSupportShape(level, basePos).getFaceShape(direction.opposite), SUPPORT_SHAPES[direction]!!, BooleanOp.AND)
    }

    override fun updateShape(
        state: BlockState,
        level: LevelReader,
        ticks: ScheduledTickAccess,
        pos: BlockPos,
        directionToNeighbour: Direction,
        neighbourPos: BlockPos,
        neighbourState: BlockState,
        random: RandomSource
    ): BlockState = if (directionToNeighbour == state[FACING] && !state.canSurvive(level, pos))
        Blocks.AIR.defaultBlockState()
    else
        super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random)

    override fun animateTick(state: BlockState, level: Level, pos: BlockPos, random: RandomSource) {
        if (level[pos.below()].isFaceSturdy(level, pos, Direction.DOWN, SupportType.FULL) || random.nextInt(10) != 0) return

        val aabb = state.getShape(level, pos).bounds()

        level.addParticle(
            ParticleTypes.DRIPPING_HONEY,
            aabb.xsize * random.nextDouble() + aabb.minX + pos.x,
            aabb.minY + pos.y - 0.05,
            aabb.zsize * random.nextDouble() + aabb.minZ + pos.z,
            0.0, 0.0, 0.0
        )
    }

    companion object {
        val FACING: EnumProperty<Direction> = DirectionalBlock.FACING

        private fun directionalLayer(direction: Direction, thickness: Double): VoxelShape = box(
            if (direction == Direction.WEST) 16.0 - thickness else 0.0,
            if (direction == Direction.DOWN) 16.0 - thickness else 0.0,
            if (direction == Direction.NORTH) 16.0 - thickness else 0.0,
            if (direction == Direction.EAST) thickness else 16.0,
            if (direction == Direction.UP) thickness else 16.0,
            if (direction == Direction.SOUTH) thickness else 16.0
        )

        val FACE_SHAPES = Direction.entries.associateWith {
            directionalLayer(it, 4.0)
        }

        val SUPPORT_SHAPES = Direction.entries.associateWith {
            directionalLayer(it.opposite, 1.0)
        }
    }
}